### ISSUE TITLE
Low-hanging fruits optimizations in Akaze feature detector

### ISSUE THREAD

**Issue by stephanemagnenat** [rust-cv/cv#67]
While working on fixing #63, I'm going through all of Akaze's source code quite deeply, and so to not loose the understanding, I'm collecting here the low-hanging fruit optimization opportunities I'm discovering. These are:
-  In `create_nonlinear_scale_space`, the `Lx` and `Ly` images could be computed in parallel, as these only access `Lsmooth` in read. This should be useful for the first evolutions when images are of a significant size.
- In `find_scale_space_extrema`, there is a O(n²) loop for testing duplicates. This could be optimized to O(n) using some form of spacing hashing.

A note here: I did not consider micro-optimizations, such as parallelizing loops within a simple image computations. If so, more could be done, but the benefit is not fully clear due to the overhead cost of spawning threads.

**Comment by vadixidav** [comment]
@stephanemagnenat I previously found that it was more efficient from a throughput perspective to perform akaze extraction on multiple frames in parallel rather than parallelizing the individual frames, which optimizes for latency. If you add rayon parallelization, please make rayon an optional dependency and gate out the rayon code when it isn't enabled, using ndarray or similar instead. For robotics, it may be preferable to use rayon for low latency odometry, whereas for mapping purposes it may be preferable to boost throughput.

**Comment by stephanemagnenat** [comment]
Of course, Rayon is an optional dependency. I always have first the normal code without rayon, and then the rayon-specific code (see #68).

For my application, which is related to AR, latency is key. I will probably also do some more detailed profiling, because one target is WASM, where rayon is generally not available.

**Comment by stephanemagnenat** [comment]
Some profiling shows that the hottest function is `imageproc::filter::separable_filter` (~50 % of time) which is not vectorized, even in release.

**Comment by stephanemagnenat** [comment]
So I have rewritten the filters by considering the memory constraints. They seem to be 15% faster than the imageproc versions, and are unit tested against these.

They are still far slower than OpenCV version, mostly because due to the lack of fast math, Rust is not able to vectorize these loops. There is an interesting discussion on the [rust user forum](https://users.rust-lang.org/t/rust-performance-help-convolution/44075/2). If we would be able to force fast math, we could expect a factor 5 of speed-up! But that is only possible on nightly! Alternatively, we can wait for portable SIMD to stabilize and use that instead. My versions are already a step in that direction.

That state of affair is a bit sad though...

**Comment by stephanemagnenat** [comment]
@vadixidav @xd009642 I'm now working on SIMD versions of filters. The first version, for horizontal filter, has 2x improvement for small kernel (7 wide) and 5x for large ones (71 wide). I believe there is still a bit of room for improvement.

This brings several questions for which I'll like your opinion:
- These functions should ideally go in `imageproc`, but there is little responsiveness from maintainers there. Should we keep them here for now?
- To develop them, I need to benchmark. To benchmark, I need to have these as public functions due to limits in criterion. This means that `horizontal_filter`, `vertical_filter`, `separable_filter` and `gaussian_kernel` become public. I do not like this so much especially in the light of their eventual inclusion in `imageproc`. But given the current reality, I do not see another path. What do you think?

Please note that I'm happy to split this PR, for example to merge the SIMD filters earlier than the use of rayon, if these are less contreversial.

**Comment by stephanemagnenat** [comment]
I think that this is ready to be merged. There are two kinds of changes:
- 1. Performance optimization on a single thread: mostly SIMD optimization of the filters (using the [wide](https://docs.rs/wide/latest/wide/) crate while waiting for the portable SIMD initiative to land) plus some minor code change (memory access reorganization).
- 2. Parallelization using Rayon, behind the `rayon` feature.

My most pressing need is 1., as I wish to deploy Akaze in an embedded WASM application. But as 2. is behind a feature flag, I think that it makes sense to merge as well.

@vadixidav and @xd009642 anything else you need from me before merging this?

**Comment by xd009642** [comment]
I don't think there's anything extra, just need to find the time to go over and review the code. One thought would be if the CI doesn't test it builds against any wasm targets with the desired features it may be worth adding that to CI to make sure nothing's added in future that breaks that.

**Comment by xd009642** [comment]
Oh there is a Cargo.toml conflict after I merged the serde PR, so that is one thing needed.

Btw I'm away this weekend so may not get to it until next week. Just trying to get through some stuff before I leave :smile:

**Comment by stephanemagnenat** [comment]
Rebased and made clippy happy.

**Comment by stephanemagnenat** [comment]
Should we also test the "rayon" feature in the CI?

**Comment by xd009642** [comment]
Ideally we should to avoid fixed only being applied to one and not the other

**Comment by stephanemagnenat** [comment]
Done in b0e7222eb7c590b5c903f2849572ed69a195aa82. The PR is ready for review.

**Comment by stephanemagnenat** [comment]
Please note that due to a limitation in github cargo actions, I could not run the test for the rayon `feature` of akaze in the `akaze` sub-directory only, so I had to run the whole cargo test again with the `--features akaze/rayon` argument. This works but is slow as unrelated parts of CV, such as SFM, get compiled again as well. If anyone knows how to streamline that, it would be welcome! Of course, later on, we might want a CV-wide rayon feature, it that case it won't matter much.

**Comment by stephanemagnenat** [comment]
I just realized that `std::time::Instant` is broken on WASM platform. I'm using it now, I will thus make it an optional feature.

**Comment by xd009642** [comment]
Sure and I'll try to find some time today or tomorrow for the (hopefully) final review before merge :+1:

**Comment by stephanemagnenat** [comment]
So the replacement doesn't have proper support for no_std, although there is a [PR for it](https://github.com/sebcrozet/instant/pull/48). I think that for simplicity, at the moment, I will disable timings when on wasm32. I'll use a macro, to easily abstract later.

**Comment by stephanemagnenat** [comment]
One question @xd009642, this PR adds the `serde` optional feature. At the moment, it is with the `serde` name, which is very common in the Rust world and also what `bitarray` uses for example.

But elsewhere in CV there is a `serde-serialize` name that itself activates `serde` and `serde` on dependencies. This is confusing because then there are two serde features: `serde` that just uses the `serde` crate and `serde-serialize` that enables `serde` on dependencies.

That is actually not necessary, as dependencies can have `serde` enabled by specifying `serde` in `[features]` like this: `serde = ["dep:serde", "bitarray/serde"]`.

So I assume that it is historical, and so far have been using just `serde`. But I could change if we want that, although I feel it makes things uselessly complicated. We could also change the rest of CV to the simpler option as an alternative.

**Comment by xd009642** [comment]
Yeah the serde thing is historical I'd prefer the simpler serde feature - and updating the crates to do the same thing (in a future PR though)

**Comment by stephanemagnenat** [comment]
Ok, good, we keep it this way then.

The PR is updated with disabled instant on wasm target. The reason I'm not quickly implementing a replacement is that it would only work in the browser, and probably not on some WASI targets. We could also have a way to provide a time-measurement function, but then additional questions arise. So probably it is better to merge as now and thing about these afterwards.

**Comment by stephanemagnenat** [comment]
I'm closing the issue as the related PR has been merged.

**Comment by kalwalt** [comment]
> Ok, good, we keep it this way then.
> 
> The PR is updated with a replacement on WASM target if the optional `web-sys` feature is enabled.
> 
> I had to do this because the WASM target is currently `wasm32-unknown-unknown` so in principle we cannot assume any specific platform support, that is why `Instant` doesn't work in the first place. The `web-sys` feature assumes we target the web and links to `web-sys` to use the browser's [Performance.now() method](https://developer.mozilla.org/en-US/docs/Web/API/Performance/now).

@stephanemagnenat i did in the past a test to export to wasm  https://github.com/kalwalt/wasm-akaze-example i will try with these recent changes and i will let you know.

**Comment by stephanemagnenat** [comment]
@kalwalt great, on my side I successfully deployed to WASM (with SIMD extension) for our application. On my phone (Galaxy S9), it takes about 2s to capture an image, find the Akaze points, match them to a template, find the homography with some outlier rejection, extract a sub-part of the image, rectify and display that part.

It is nowhere native performance but it is usable.
